File: leetcode/C++Solutions/valid-sudoku.cpp

#include <vector>
#include <unordered_set>

using namespace std;

class Solution {
public:
    bool isValidSudoku(vector<vector<char>>& board) {
        for (int i = 0; i < 9; ++i) {
            unordered_set<char> row;
            unordered_set<char> col;
            unordered_set<char> box;
            for (int j = 0; j < 9; ++j) {
                // Check rows
                if (board[i][j] != '.') {
                    if (row.find(board[i][j]) != row.end()) {
                        return false;
                    }
                    row.insert(board[i][j]);
                }
                // Check columns
                if (board[j][i] != '.') {
                    if (col.find(board[j][i]) != col.end()) {
                        return false;
                    }
                    col.insert(board[j][i]);
                }
                // Check sub-boxes
                int box_row = 3 * (i / 3) + j / 3;
                int box_col = 3 * (i % 3) + j % 3;
                if (board[box_row][box_col] != '.') {
                    if (box.find(board[box_row][box_col]) != box.end()) {
                        return false;
                    }
                    box.insert(board[box_row][box_col]);
                }
            }
        }
        return true;
    }
};

File: leetcode/pythonSolutions/valid-sudoku.py
# class Solution:
#     def isValidSudoku(self, board: List[List[str]]) -> bool:
        
#         cols = collections.defaultdict(set)
#         rows = collections.defaultdict(set)
#         squares = collections.defaultdict(set)

#         for r in range(9):
#             for c in range(9):
#                 if board[r][c] == ".":
#                     continue
#                 if (board[r][c] in rows[r] or board[r][c] in cols[c] or board[r][c] in squares[( r // 3, c // 3 )]):
#                     return False
#                 cols[c].add(board[r][c])
#                 rows[r].add(board[r][c])
#                 squares[( r // 3, c // 3 )].add(board[r][c])
#         return True

# the above answer is from neetcode, the difference between the two solutions is that neetcode checks if the element is present or not in the square key instead of iterating through an entire sub square. 

class Solution:
    def isValidSudoku(self, board: List[List[str]]) -> bool:
    
        for i in range(9):
            row = set()
            col = set()
            box = set()
            for j in range(9):
                # Check rows
                if board[i][j] != ".":
                    if board[i][j] in row:
                        return False
                    row.add(board[i][j])
                # Check columns
                if board[j][i] != ".":
                    if board[j][i] in col:
                        return False
                    col.add(board[j][i])
                # Check sub-boxes
                box_row = 3 * (i // 3) + j // 3
                box_col = 3 * (i % 3) + j % 3
                if board[box_row][box_col] != ".":
                    if board[box_row][box_col] in box:
                        return False
                    box.add(board[box_row][box_col])
        return True



